### PR TITLE
CERT 7660 Add CoinGecko + Price Feed API Refactor

### PR DESCRIPTION
**Pull request by nivcertora**
https://certora.atlassian.net/browse/CERT-7660

### **Pull Request Summary: Integrate CoinGecko as a New Price Feed Provider and Refactor Existing APIs**

**Overview:**
This pull request introduces CoinGecko as an additional price feed provider within the Quorum project. It also refactors existing price feed APIs to support address-specific data retrieval, enhances the `PriceFeedData` model, and updates configuration and documentation to accommodate the new integration.

---

**Key Changes:**

1. **Addition of `CoinGeckoAPI` Provider:**
   - **File Added:** `Quorum/apis/price_feeds/coingecko_api.py`
   - **Description:** Introduces the `CoinGeckoAPI` class, which interacts with the CoinGecko API to fetch and validate token price feed data based on contract addresses.
   - **Features:**
     - **Platform Mapping:** Establishes a mapping between Quorum's supported chains and CoinGecko's platforms.
     - **Singleton Pattern:** Ensures a single instance of `CoinGeckoAPI` using the `@singleton` decorator.
     - **Data Retrieval:** Implements the `_get_price_feed_info` method to fetch and process price feed data for specific addresses.

2. **Refactoring of Existing Providers (`ChainLinkAPI` and `ChronicleAPI`):**
   - **Files Modified:**
     - `Quorum/apis/price_feeds/chainlink_api.py`
     - `Quorum/apis/price_feeds/chronicle_api.py`
   - **Description:** Updates both `ChainLinkAPI` and `ChronicleAPI` to support address-specific data retrieval instead of fetching all feeds for a chain.
   - **Changes:**
     - **Method Renaming:** Changes `_get_price_feeds_info` to `_get_price_feed_info`, accepting both `chain` and `address` as parameters.
     - **Caching Mechanism:** Adjusts caching to store and retrieve `PriceFeedData` instances based on contract addresses.
     - **Data Processing:** Modifies data parsing to handle individual address lookups efficiently.

3. **Enhancements to `PriceFeedData` Model:**
   - **File Modified:** `Quorum/apis/price_feeds/price_feed_utils.py`
   - **Description:** Enhances the `PriceFeedData` model for better clarity and functionality.
   - **Improvements:**
     - **Decimals Field:** Adds a new `decimals: Optional[int]` field to accurately represent token decimals.
     - **Custom String Representation:** Implements the `__str__` method to provide a clear and coherent string representation of `PriceFeedData` instances, handling optional fields gracefully.

4. **Configuration Updates:**
   - **Files Modified:**
     - `Quorum/utils/config_loader.py`
     - `Quorum/ground_truth.json`
   - **Description:** Updates configuration files to include the new `CoinGeckoAPI` provider.
   - **Changes:**
     - **`config_loader.py`:** Incorporates `CoinGeckoAPI` into the list of supported price feed providers.
     - **`ground_truth.json`:** Adds `"CoinGecko"` to the `"price_feed_providers"` array to enable its usage.

### CODE REVIEW

File: Quorum/apis/git_api/git_manager.py [hunk 1]
```diff
@@ -36,22 +36,16 @@ def _load_repos_from_file(self, gt_config: dict[str, any]) -> tuple[dict[str, st
...
                 The first dictionary contains the repos to diff against. The second dictionary is the verification repo.
         """
-        # Normalize the customer name to handle case differences
-        normalized_customer = self.customer.lower()
-        customer_repos = next((repos for key, repos in gt_config.items() if key.lower() == normalized_customer), None)
-        if customer_repos is None:
-            return {}, {}
-        
-        repos = {Path(r).stem: r for r in customer_repos["dev_repos"]}
+        repos = {Path(r).stem: r for r in gt_config["dev_repos"]}
```
Comment (yoav-el-certora) [inline Quorum/apis/git_api/git_manager.py:45]: This is a bug fix?

File: Quorum/apis/git_api/git_manager.py [hunk 2]
```diff
@@ -36,22 +36,16 @@ def _load_repos_from_file(self, gt_config: dict[str, any]) -> tuple[dict[str, st
...
 
         Returns:
             tuple[dict[str, str], dict[str, str]]: 2 dictionaries mapping repository names to their URLs.
                 The first dictionary contains the repos to diff against. The second dictionary is the verification repo.
         """
-        # Normalize the customer name to handle case differences
-        normalized_customer = self.customer.lower()
-        customer_repos = next((repos for key, repos in gt_config.items() if key.lower() == normalized_customer), None)
-        if customer_repos is None:
-            return {}, {}
```
Comment (yoav-el-certora) [inline Quorum/apis/git_api/git_manager.py:49]: Why did we remove this?

**Review by yoav-el-certora — COMMENTED** [review]
Looks fine in general, can comment on some improvements in logic but will wait for @liav-certora deep review.
Next time, lets split both PR's as this implement 2 different tasks...

Additionally, please add relevant information to README file.

File: Quorum/apis/git_api/git_manager.py (hunk 2 above)
Comment (nivcertora) [reply Quorum/apis/git_api/git_manager.py:49]: That was a bug that prevented from the git repo to update

File: Quorum/apis/price_feeds/coingecko_api.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        """
+        platform = self.CHAIN_TO_COINGECKO_PLATFORM_MAP.get(chain)
+        if not platform:
+            return None
+        
+        url = self.COINGECKO_API_URL.format(platform=platform, address=address)
+        response = self.session.get(url)
+        if response.status_code != 200:
+            return None
+        response.raise_for_status()
```
Comment (liav-certora) [inline Quorum/apis/price_feeds/coingecko_api.py:49]: Redundant due to check above.

File: Quorum/apis/price_feeds/coingecko_api.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        data: dict = response.json()
+        if not data:
+            return None
+        detail_platforms: dict = data.get('detail_platforms')
+        if not detail_platforms:
+            return None
+        platform_details = detail_platforms.get(platform)
+        if not platform_details:
+            return None
```
Comment (liav-certora) [inline Quorum/apis/price_feeds/coingecko_api.py:58]: Would appreciate better naming for the variables here. "detail_platforms" and "platform_details" is pretty confusing.

File: Quorum/apis/price_feeds/price_feed_utils.py [hunk 5]
```diff
@@ -40,32 +60,36 @@ def __init__(self):
         self.cache_dir = Path(__file__).parent / "cache" / self.get_name().value
         self.cache_dir.mkdir(parents=True, exist_ok=True)
         self.session = requests.Session()
-        self.memory: dict[Chain, dict[str, PriceFeedData]] = {}
+        self.memory: dict[str, PriceFeedData] = {}
```
Comment (liav-certora) [inline Quorum/apis/price_feeds/price_feed_utils.py:63]: I'm not sure if this change is very safe... What if my some unlucky chance you'll have the same address on 2 chains? I do see you divide the actual cache on the file system to chain -> address but not in code.

File: Quorum/apis/price_feeds/price_feed_utils.py [hunk 6]
```diff
@@ -40,32 +60,36 @@ def __init__(self):
...
-                self.memory[chain] = self._get_price_feeds_info(chain)
+            if address not in self.memory:
+                self.memory[address] = self._get_price_feed_info(chain, address)
+                if not self.memory[address]:
+                    return None
+                cache_file.parent.mkdir(parents=True, exist_ok=True)
                 with open(cache_file, 'w') as file:
-                    json.dump({addr: feed.dict() for addr, feed in self.memory[chain].items()}, file)
-        return self.memory[chain]
+                    json.dump(self.memory[address].model_dump(mode="json"), file)
```
Comment (liav-certora) [inline Quorum/apis/price_feeds/price_feed_utils.py:88]:
```suggestion
                    json.dump(self.memory[address].model_dump(mode="json"), file, indent=4)
```
For readability if we'll ever need it.

File: Quorum/apis/price_feeds/price_feed_utils.py (hunk 5 above)
Comment (nivcertora) [reply Quorum/apis/price_feeds/price_feed_utils.py:63]: Lol, its not going to happen dont worry

**Review by liav-certora — APPROVED** [review]
looks good

**Pull request merged**